### ISSUE TITLE
Feature Request: A way to redefine mesh specifications within a shard_map

### ISSUE THREAD

**Issue by chaserileyroberts** [google/jax#18537]
There are situations where one way of labeling the same mesh can make certain communication patterns easier to code than others. However, there isn't really one mesh labeling strategy that is consistently easy to use across the space of all possible communication patterns. Currently, if a user or library developer wants to develop advanced communication patterns via `shard_map`, they'll be forced to either 

a) Force end users to use only a predefined set of axis labels for certain libraries. If these are not consistent between libraries, then you can not use those libraries together.
b) User explicitly passes axis labels for all ops. Both library developers and end users are forced to lug around passed axes names throughout their code. It's a large burden, and very fragile to change.

This is what I propose instead: A new method `mesh_remap(fn: Callable, mesh: Mesh)` that can be used within any normal `shard_map`ed method. The basic use case looks like this:

```python
devices = jax.devices().reshape((16, 16))
mesh_2d = Mesh(devices.reshape((16, 16)), ['x', 'y'])
mesh_3d = mesh(devices.reshape((8, 2, 16), ['a', 'b', 'c'])


@partial(mesh_remap, mesh=mesh_3d)
def gather_middle(arr):
  # arr is still the same 2D tile from f.
  # Nothing has changed. Not the shape nor the device
  # placement

  # We gather against he middle 'b' mesh axis.
  res == all_gather(arr, 'b', tiled=True)
  assert arr.shape[0] == res.shape[0] // 2
  return res

@partial(shard_map, mesh=mesh_2D, in_spec=P('x', 'y'), out_spec=P('x', 'y'))
def f(arr):
  # arr is a 2D tile where the first dimension is doubled. 
  return gather_middle(arr)
```

Here, all we are doing when calling `gather_middle` is redefining how the mesh is labeled. `arr` is not communicated in any way until the `all_gather(...)` call. In this case, since we are gathering on the new `'b'` mesh axis, this all gather definition would not have been possible with the previous `mesh_2d` axes.

The same code could potentially be used instead with a `with` block. Since we do not need to define input or output specification, wrapping a function is not required. We are only changing the global state of the mesh, so a with clause should theoretically suffice.

```python
@partial(shard_map, mesh=mesh_2D, in_spec=P('x', 'y'), out_spec=P('x', 'y'))
def f(arr):
  with mesh_remap(mesh_3D):
    return all_gather(arr, 'b', tiled=True)
```

Which is best from a user or jaxpr standpoint is something that should likely be discussed.

Alternative considerations:

1. Simply use really high dimensional meshes and and use axes tuples (for example: `all_gather(x,  axes=('i', 'j'))`) for larger groupings.

In practice this solution has been what we've been doing. The above example could've instead been implemented like this 

```python
@partial(shard_map, mesh=mesh_3D, in_spec=P(('a', 'b'), 'c'), out_spec=P(('a', 'b'), 'c'))
def f(arr):
  return gather_middle(arr)
```

And this works plus or minus a few edge case bugs. This is enough for small benchmarks, but it's limited in flexibility, and you'll be forced to deal with many multiple more axes at any given time than is convenient to deal with. This solution also limits the communication strides you can use to what is was possible with your initial mesh definition. Library writers are also forced to deal with user defined mesh labels, and it can become a headache when doing automated communication pattern optimizations.

2. Use independent shard_maps and pipe the results between each method.

Can work, but is fragile. If the partition spec / mesh alignment is not consistent you may unintentionally be inserting additional comms to match input signatures.
